### ISSUE TITLE
Need to encode PID in writePIDFile in common.py

### ISSUE THREAD

**Issue by elarsonneur** [DataBiosphere/toil#2523]
Hi,
This error happens when running cwltoil in python3.6 and not in python2.7: 

```
cwltoil [...]
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/cwltoil", line 10, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python3.6/site-packages/toil/cwl/cwltoil.py", line 1220, in main
    outobj = toil.start(wf1)
  File "/usr/local/miniconda/lib/python3.6/site-packages/toil/common.py", line 758, in start
    self.writePIDFile()
  File "/usr/local/miniconda/lib/python3.6/site-packages/toil/common.py", line 1086, in writePIDFile
    f.write(str(os.getpid()))
TypeError: a bytes-like object is required, not 'str'
```

cwltoil release : 3.18.0
python version : 3.6

Elise

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-53)
┆Issue Number: TOIL-53

**Comment by illusional** [comment]
Hey @elarsonneur, I ran into the same problem as you, however it seems Toil [only supports Python 2.7](https://toil.readthedocs.io/en/latest/gettingStarted/install.html#preparing-your-python-runtime-environment): 

> Toil currently supports only Python 2.7 and requires a virtualenv to be active to install.

There seemed to be some work going into making it compatible with Python 3+ (https://github.com/DataBiosphere/toil/issues/1780) and the [Toil roadmap](https://github.com/DataBiosphere/toil/wiki/Roadmap) has this feature request as _Near Term (In progress)_. I'm also interested in Python3, so I've pinged some people on #1780 and I'll update this comment if I find anything else out.

**Comment by elarsonneur** [comment]
Thanks @illusional to remind me there is no full support for Python3.
In the [release3.18.0](https://github.com/DataBiosphere/toil/releases), they mentionned:
> Large python 2/3 compatibility changes (+ python 3 compatible mesos library: pymesos)

**Comment by adamnovak** [comment]
We *should* support Python 3 now. Unfortunately, in practice we're still squishing bugs, one of which is that line in the documentation that says we don't support Python 3.

**Comment by drkennetz** [comment]
I have added some logic to deal with this, but in a different place.

I was encountering the same error in `lsf.py`
```
        for line in p.stdout:
                split_items = line.strip().split()
#logic          items = [item.decode() for item in split_items if isinstance(item, bytes)]
                for value in items:
                    if isinstance(value, bytes):
                        value = value.decode()
                if len(items) < num_columns:
                        RuntimeError("lshosts output has a varying number of "
                                     "columns")
#logic      if items[cpu_index] != '-' and int(items[cpu_index]) > int(maxCPU):
                        maxCPU = items[cpu_index]
                if (items[mem_index] != '-' and
                        MemoryString(items[mem_index]) > maxMEM):
                    maxMEM = MemoryString(items[mem_index])
```

And added the logic at each place to deal with this. It seems toil assumes that no system input file will be compressed. Perhaps a decorator should be added before these functions that require system inputs to convert them to string if they are bytes.

**Comment by adamnovak** [comment]
OK @drkennetz, I've opened #2724 to track your suggested change. If you want to submit it in the form of a PR, we can take that more or less right away. Otherwise we'll get to it eventually.

**Comment by jeffrey856** [comment]
It seems fixed already.

**Comment by drkennetz** [comment]
@jeffrey856 some bugs in `lsf.py` are fixed by my PR in bc5866f4894e9f034bdc2baea5bdf0d6ebacf42a
